bump(deps): upgrade rust `bitcoin` to `0.32.0`, `miniscript` to `0.12.0` and others

oleonardolima opened the pull request:
fixes bitcoindevkit/bdk#1422
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

This PR focuses on upgrading the `rust-bitcoin` and `miniscript` versions, to `0.32.0` and `0.12.0`, respectively. It also bumps the versions of other BDK ecosystem crates that also rely on both `rust-bitcoin` and `miniscript`, being: 

- electrum-client https://github.com/bitcoindevkit/rust-electrum-client/pull/133
- esplora-client https://github.com/bitcoindevkit/rust-esplora-client/pull/85
- hwi https://github.com/bitcoindevkit/rust-hwi/pull/99

<ins>I structured the PR in multiple commits, with closed scope, being one for each BDK crate being upgraded, and one for each kind of fix and upgrade required, it seems like a lot of commits (**that should be squashed before merging**), but I think it'll make it easier during review phase.</ins>

In summary I can already mention some of the main changes:
- using `compute_txid()` instead of deprecated `txid()`
- using `minimal_non_dust()` instead of `dust_value()`
- using the renamed `signature` and `sighash_type` fields 
- using proper `sighash::P2wpkhError`,  `sighash::TaprootError` instead of older `sighash::Error`
- conversion from `Network` to new expected `NetworkKind` bitcoindevkit/bdk_wallet#94 
- conversion from the new `Weight` type to current expected `usize` bitcoindevkit/bdk#1466 
- using `.into()` to convert from AbsLockTime and `RelLockTime` to `absolute::LockTime` and `relative::LockTime`
- using Message::from_digest() instead of relying on deprecated `ThirtyTwoByteHash` trait.
- updating the miniscript policy and dsl to proper expect and handle new `Threshold` type, instead of the previous two parameters.

<!-- Describe the purpose of this PR, what's being adding and/or fixed -->

### Notes to the reviewers
<ins>Again, I structured the PR in multiple commits, with closed scope, being one for each BDK crate being upgraded, and one for each kind of fix and upgrade required, it seems like a lot of commits (**that should be squashed before merging**), but I think it'll make it easier during review phase.</ins>

It should definitely be carefully reviewed, especially the last commits for the wallet crate scope, the ones with the semantic `fix(wallet)`.

I would also appreciate if @tcharding reviewed it, as he gave a try in the past (#1400 ), and I relied on some of it for the  policy part of it, other rust-bitcoin maintainers reviews are a definitely welcome 😁 

<!-- In this section you can include notes directed to the reviewers, like explaining why some parts
of the PR were done in a specific way -->

### Changelog notice
- Use `compute_txid()` instead of deprecated `txid()`
- Use `minimal_non_dust()` instead of `dust_value()`
- Use `signature` and `sighash_type` fields, instead of previous `sig` and `hash_type` 
- Use `sighash::P2wpkhError`,  and `sighash::TaprootError` instead of older `sighash::Error`
- Converts from `Network` to `NetworkKind`, where expected
- Converts from `Weight` type to current used `usize`
- Use `.into()` to convert from `AbsLockTime` and `RelLockTime` to `absolute::LockTime` and `relative::LockTime`
- Remove use of  deprecated `ThirtyTwoByteHash` trait, use `Message::from_digest()` 
- Update the miniscript policy and dsl macros to proper expect and handle new `Threshold` type, instead of the previous two parameters.

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

oleonardolima committed:
deps(bdk): bump `bitcoin` to `0.32.0`, miniscript to `12.0.0`

deps(chain): bump `bitcoin` to `0.32.0`, miniscript to `12.0.0`

fix(chain): use `minimal_non_dust()` instead of `dust_value()`

fix(chain): use `compute_txid()` instead of `txid`

deps(testenv): bump `electrsd` to `0.28.0`

deps(electrum): bump `electrum-client` to `0.20.0`

fix(electrum): use `compute_txid()` instead of `txid`

deps(esplora): bump `esplora-client` to `0.8.0`

deps(bitcoind_rpc): bump `bitcoin` to `0.32.0`, `bitcoincore-rpc` to
`0.19.0`

fix(bitcoind_rpc): use `compute_txid()` instead of `txid`

fix(nursery/tmp_plan): use proper `sighash` errors, and fix the expected
`Signature` fields

fix(sqlite): use `compute_txid()` instead of `txid`

deps(hwi): bump `hwi` to `0.9.0`

deps(wallet): bump `bitcoin` to `0.32.0`, miniscript to `12.0.0`

fix(wallet): use `compute_txid()` and `minimal_non_dust()`

- update to use `compute_txid()` instead of deprecated `txid()`
- update to use `minimal_non_dust()` instead of `dust_value()`
- remove unused `bitcoin::hex::FromHex`.

fix(wallet): uses `.into` conversion on `Network` for `NetworkKind`

- uses `.into()` when appropriate, otherwise use the explicit
  `NetworkKind`, and it's `.is_mainnet()` method.

fix(wallet): add P2wpkh, Taproot, InputsIndex errors to `SignerError`

fix(wallet): fields on taproot, and ecdsa `Signature` structure

fix(wallet/wallet): convert `Weight` to `usize` for now

- converts the `bitcoin-units::Weight` type to `usize` with help of
  `to_wu()` method.
- it should be updated/refactored in the future to handle the `Weight`
  type throughout the code instead of current `usize`, only converting
  it for now.
- allows the usage of deprecated `is_provably_unspendable()`, needs
  further discussion if suggested `is_op_return` is suitable.
- update the expect field to `signature`, as it was renamed from `sig`.

fix(wallet/wallet): use `is_op_return` instead of
`is_provably_unspendable`

fix(wallet/wallet): use `relative::Locktime` instead of `Sequence`

fix(wallet/descriptor): use `ParsePublicKeyError`

fix(wallet/descriptor): use `.into()` to convert from `AbsLockTime` and
`RelLockTime` to `absolute::LockTime` and `relative::LockTime`

fix(wallet/wallet): use `Message::from_digest()` instead of relying on
deprecated `ThirtyTwoByteHash` trait.

fix(wallet/descriptor+wallet): expect `Threshold` type, and handle it
internally

fix(wallet/wallet): remove `0x` prefix from expected `TxId` display

fix(examples): use `compute_txid()` instead of `txid`

fix(ci): remove usage of `bitcoin/no-std` feature

- remove comment: `# The `no-std` feature it's implied when the `std` feature is disabled.`

oleonardolima commented:
NOTE: related to the mentioned todo's on the description, I'll open a new PR to add more scope/context to this one with all the changes required due to the new `NetworkKind` and new `Weight`.

ValuedMammal commented on crates/wallet/src/descriptor/error.rs:
```diff
@@ -37,7 +37,7 @@ pub enum Error {
     Base58(bitcoin::base58::Error),
     /// Key-related error
-    Pk(bitcoin::key::Error),
+    Pk(bitcoin::key::ParsePublicKeyError),
```
I'm undecided if we need this `Pk` variant

oleonardolima replied:
What are your thoughts about it?

ValuedMammal replied:
afaik it's not being used anywhere

ValuedMammal commented on crates/wallet/src/descriptor/policy.rs:
```diff
@@ -952,11 +999,14 @@ impl<Ctx: ScriptContext + 'static> ExtractPolicy for Miniscript<DescriptorPublic
             }
             Terminal::Older(value) => {
-                let mut policy: Policy = SatisfiableItem::RelativeTimelock { value: *value }.into();
+                let mut policy: Policy = SatisfiableItem::RelativeTimelock {
+                    value: (*value).into(),
+                }
+                .into();
```
What do you think about having `SatisfiableItem::RelativeTimelock` hold a value of type `relative::LockTime` instead of `Sequence` ?

notmandatory replied:
Keep in mind that this module will likely be deprecated in the next major release (2.0) in favor of the new miniscript [plan module](https://docs.rs/miniscript/latest/miniscript/plan/index.html). Unless we're fixing a bug it may not be worth the refactoring to improve the API.

oleonardolima replied:
@ValuedMammal I think that's a great suggestion, but I'm now unsure if it's worth it considering what @notmandatory, I'll take a look if it's only a few LOC/changes that would be needed for the improvement.

oleonardolima replied:
@ValuedMammal I updated the variant, it was trivial. I think the most of work would be updating `Condition` (below), but I don't its work as @notmandatory  mentioned 🤔  https://github.com/bitcoindevkit/bdk/blob/473ef9714f642ee2ad393541763881cf69132262/crates/wallet/src/descriptor/policy.rs#L446-L456

ValuedMammal replied:
`Condition` should be fine as it is now

ValuedMammal commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1502,10 +1502,12 @@ impl Wallet {
             if !params.allow_dust
                 && value.is_dust(script_pubkey)
                 && !script_pubkey.is_provably_unspendable()
+            // FIXME: (@leonardo) I think it needs further discussion, should we rely on the suggested `is_op_return` ?
```
Yes I'm in favor of using `is_op_return`

notmandatory replied:
Agreed, we should use `is_op_return()`, especially since `is_provably_unspendable()` is now deprecated.

oleonardolima replied:
Great, I'll fix it to use it then.

ValuedMammal commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1704,7 +1706,9 @@ impl Wallet {
         {
-            return Err(BuildFeeBumpError::IrreplaceableTransaction(tx.txid()));
+            return Err(BuildFeeBumpError::IrreplaceableTransaction(
+                tx.compute_txid(),
```
Also, L1671 there's an old `.txid()` in a doc comment

notmandatory replied:
A few other spots in doc code that need to have tx.txid() fixed:
chain, tx_data_traits.rs: L46, L61, L79
wallet, wallet/mod.rs: L1671

oleonardolima replied:
Thanks, sorry for the oversight, it's probably because `rust-analyzer` does not handle these due to being under feature/docs codes, and it didn't show up as failure. I'll handle these.

ValuedMammal commented on crates/wallet/src/wallet/signer.rs:
```diff
@@ -548,21 +566,24 @@ fn sign_psbt_ecdsa(
 ) {
-    let msg = &Message::from(hash);
-    let sig = if allow_grinding {
+    let msg = &Message::from_digest(hash.to_byte_array());
```
We could also just pass a `&Message` to the function `sign_psbt_ecdsa`

oleonardolima replied:
@ValuedMammal Wouldn't it add more complexity/verbosity ? 🤔

ValuedMammal replied:
I'm thinking this module might eventually look more like in #1424, but if this works for now then that's great too

ValuedMammal commented on crates/wallet/src/wallet/signer.rs:
```diff
@@ -158,8 +158,12 @@ pub enum SignerError {
+    /// Error while computing the hash to sign a Taproot input.
+    SighashTaproot(sighash::TaprootError),
+    /// Error while computing the hash, out of bounds access on the transaction inputs.
+    InputsIndexError(transaction::InputsIndexError),
```
📌  Considering we already have a `InputIndexOutOfRange` error, I recommend consolidating these into one variant if possible.

oleonardolima replied:
@ValuedMammal Oh, thanks! Yes, that's much better, I'll fix it.

oleonardolima replied:
@ValuedMammal I did a deep-dive in rust-bitcoin, and now I'm wondering if we should have two variants, one for Psbt and another one for Transaction, as they have two in rust-bitcoin crate:
- bitcoin::psbt::IndexOutOfBoundsError
- bitcoin::transaction::InputsIndexError & bitcoin::transaction::OutputsIndexError

oleonardolima replied:
I updated it to be `TxInputsIndexError` instead, I also think for the current `InputIndexOutOfRange` it should be `PsbtInputIndexOutOfRange(psbt::IndexOutOfBoundsError)` instead, let me know what you think.

ValuedMammal replied:
A simple solution would be to make `SignerError::InputIndexOutOfRange` into a struct with fields for `index` and `length` (essentially reimplementing [`IndexOutOfBoundsError`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.IndexOutOfBoundsError.html)) and then we have

```rust
impl From<transaction::InputsIndexError> for SignerError {}
```

similar to what you have here, but it shouldn't be a blocker to this PR. Accessing the outputs of a tx or psbt doesn't appear to be a concern in this module.

ValuedMammal commented:
@tcharding Do the changes related to the new `Threshold` still look ok to you?

notmandatory commented on crates/chain/src/descriptor_ext.rs:
```diff
@@ -31,7 +31,7 @@ impl DescriptorExt for Descriptor<DescriptorPublicKey> {
             .script_pubkey()
-            .dust_value()
+            .minimal_non_dust()
             .to_sat()
```
📌 since this is a public API should we change the `DescriptorExt` trait to make this an `Amount` instead of u64?

oleonardolima replied:
@notmandatory Yes, I do agree. I'll update it as well.

oleonardolima replied:
@notmandatory I'll address it in another PR, it ends up breaking and requiring changes on the `CoinSelectorOpt`, and probably will also raise some discussions, which is outside the scope of this PR.

notmandatory commented on crates/esplora/Cargo.toml:
```diff
@@ -13,21 +13,21 @@ readme = "README.md"
-# use these dependencies if you need to enable their /no-std features
-bitcoin = { version = "0.31.0", optional = true, default-features = false }
-miniscript = { version = "11.0.0", optional = true, default-features = false }
+# For no-std, remember it's already implied when not enabling the `std` feature.
```
⛏️ nit, I think this description would be more clearly stated (here and other files) as:
```suggestion
# The `no-std` feature it's implied when the `std` feature is disabled.
```

oleonardolima replied:
@notmandatory Great, thanks for the suggestion.

notmandatory commented on crates/wallet/src/types.rs:
```diff
@@ -72,7 +72,7 @@ pub struct WeightedUtxo {
     ///
     /// [weight units]: https://en.bitcoin.it/wiki/Weight_units
-    pub satisfaction_weight: usize,
+    pub satisfaction_weight: usize, // FIXME: (@leonardo) I think it should expect the new bitcoin-units `Weight` type instead of usize.
```
👍 agreed, now that there's a `Weight` type we should use it. If you think it's too big a change to add to this PR feel free to make an issue to fix it in a future PR.

oleonardolima replied:
@notmandatory I wasn't so sure about, but I think it adds unrelated scope/context, I'm addressing this one in  https://github.com/bitcoindevkit/bdk/pull/1468

notmandatory reviewed (commented):
I have a couple small suggestions but otherwise this looks almost ready. Great work!

I also created #1469 to remove `tmp_plan`, should it be done with this PR?

tcharding commented on crates/esplora/Cargo.toml:
```diff
@@ -13,21 +13,21 @@ readme = "README.md"
 [features]
 default = ["std", "async-https", "blocking-https-rustls"]
-std = ["bdk_chain/std"]
+std = ["bdk_chain/std", "miniscript?/std"]
```
What does the `?` do in this line?

ValuedMammal replied:
Basically, only enable the std feature if something else already enabled the miniscript dep.

https://doc.rust-lang.org/cargo/reference/features.html#dependency-features

tcharding replied:
Oh sweet, TIL. (Doesn't help us in `rust-bitcoin` yet though because it needs Rust 1.60)

evanlinjin replied:
TIL +1

tcharding commented on crates/esplora/Cargo.toml:
```diff
@@ -13,21 +13,21 @@ readme = "README.md"
-# use these dependencies if you need to enable their /no-std features
-bitcoin = { version = "0.31.0", optional = true, default-features = false }
-miniscript = { version = "11.0.0", optional = true, default-features = false }
+# The `no-std` feature it's implied when the `std` feature is disabled.
```
This comment is not really correct, there is no `no-std` feature anymore in `rust-bitcoin`. Its in another place also.

oleonardolima replied:
Good point! I think it's best to just remove the comments altogether then.
fyi @notmandatory

tcharding commented:
> @tcharding Do the changes related to the new `Threshold` still look ok to you?

Handwavey response, if it builds ship it.

Take everything I saw with a pinch of salt, I'm not that good with `miniscript`.

I'm not sure, I don't exactly know what you are aiming for:

1. Just make it build - then CI can answer that.
2. Make it build and not be obviously wrong - seems like the usage is ok to me.
3. Fully utilise `miniscript`'s new `Threshold` type - definitely currently not doing this.

I didn't grok the `fragment` macro totally but was surprised to see `expect` called if inputs to the macro are invalid thresholds.  Passing the closure is a bit hacky but I totally feel your pain, patching macros is an epic PITA.

Sorry I can't be much more help than that.

ValuedMammal commented:
The summary of changes in the description certainly helps, thanks for that. In terms of review, the individual commits are less useful since realistically this should be a single commit - the one that finally builds with tests passing.

ValuedMammal commented:
> I'm not sure, I don't exactly know what you are aiming for:
> 
>     1. Just make it build - then CI can answer that.
> 
>     2. Make it build and not be obviously wrong - seems like the usage is ok to me.
> 
>     3. Fully utilise `miniscript`'s new `Threshold` type - definitely currently not doing this.

Mostly number 2 for now, ha. All good, appreciate you @tcharding !

ValuedMammal commented on crates/wallet/src/wallet/signer.rs:
```diff
@@ -159,8 +159,14 @@ pub enum SignerError {
+    /// Error while computing the hash, out of bounds access on the transaction inputs.
+    TxInputsIndexError(transaction::InputsIndexError),
+    /// Error while computing the hash, out of bounds access on the transaction outputs.
+    TxOutputsIndexError(transaction::OutputsIndexError),
```
I'm not sure the outputs are relevant to computing the sighash

```suggestion
```

oleonardolima commented:
> The summary of changes in the description certainly helps, thanks for that. In terms of review, the individual commits are less useful since realistically this should be a single commit - the one that finally builds with tests passing.

Nice, thanks for the feedback! I'll squash the commits into a single one.

ValuedMammal commented:
ACK 2a4564097bd70905ed9fa21b92cdf074a72f878e

evanlinjin commented on crates/wallet/src/descriptor/policy.rs:
```diff
@@ -644,6 +640,57 @@ impl Policy {
         Ok(Some(policy))
     }
 
+    fn make_multi_a<Ctx: ScriptContext + 'static>(
```
~~Thanks for the attempt to rewrite what is here! I would not bother though, since we want to replace this `policy` stuff with `miniscript`'s planning module.~~

Edit: okay I'm trying to figure out why the policy will be different here?

evanlinjin replied:
Now I understand. The only difference is when constructing the `Threshold`, we enforce different max `n` (of the `k-of-on` theshold) for different script constructions.

Can we use a const generic instead of duplicating code?

```rust
    fn make_multi<Ctx: ScriptContext + 'static, const MAX: usize>(
        threshold: &Threshold<DescriptorPublicKey, MAX>,
        signers: &SignersContainer,
        build_sat: BuildSatisfaction,
        sorted: bool,
        secp: &SecpCtx,
    ) -> Result<Option<Policy>, PolicyError> { todo!() }
```

evanlinjin replied:
Due to time constraints I'm going to do this and push a commit on top.

evanlinjin replied:
Done in 11200810d048abb8deb3c7961ca93c68011d41fd

oleonardolima replied:
Great! That's much better, thanks for the suggestion and for addressing this @evanlinjin!

evanlinjin commented on crates/wallet/src/wallet/tx_builder.rs:
```diff
@@ -295,7 +295,9 @@ impl<'a, Cs> TxBuilder<'a, Cs> {
-                let satisfaction_weight = descriptor.max_weight_to_satisfy().unwrap();
+
+                let satisfaction_weight =
+                    descriptor.max_weight_to_satisfy().unwrap().to_wu() as usize;
```
Okay we need to create a ticket for what we should do when a caller creates a wallet with descriptors that are insatiable. Panicing with no clear reason when they try construct a tx is not the right behavior imo. Out of scope for this PR.

oleonardolima replied:
I agree, I tried to solve that problem in #1468 by propagating the error, but it needs further discussion.

oleonardolima replied:
I moved #1468 to ready to review now, we can move this discussion there then.

apoelstra replied:
Maybe upstream could add a field to the `Params` structure which would refuse to parse un-satiable descriptors. Could you file an issue in rust-miniscript asking for strategies to handle the error return of max_weight_to_satisfy?

Alternately you can just use `unwrap_or(usize::MAX)` or something silly.

oleonardolima replied:
Yes, I did it https://github.com/rust-bitcoin/rust-miniscript/issues/695

evanlinjin committed:
chore(wallet): rm dup code

Methods `make_multi` and `make_multi_a` were essentially the same thing
except for a different const generic param on `Threshold`. So we rm
`make_multi_a` and put a const generic param on `make_multi`.

notmandatory reviewed (approved):
ACK 11200810d048abb8deb3c7961ca93c68011d41fd
